[MRG] Adding bounds checking for `--scaled` and `--num` in `sourmash sketch`

keyabarve opened the pull request:
Fixes #1708

- Add checks for the values of `scaled` and `num` in `sourmash sketch`. 
- num values should always be > 0. if they are unreasonable, say > 50,000 or < 50, we should emit a warning.
- scaled values should always be > 0. if they are unreasonable, say > 1e6 or < 100, we should emit a warning.
- Add tests to check this behavior.

keyabarve committed:
Initial commit

keyabarve committed:
Removed comment

keyabarve commented:
@ctb I'm not being able to figure out how and where to get started. Could you provide some hints?

ctb commented:
hi @keyabarve, have you found the param string parsing code?

keyabarve commented:
Yes, I found it in `src/sourmash/command_sketch.py`. This is the function: `def _parse_params_str(params_str):`.

ctb commented:
ok. So that's where current value checking is happening, and where any new value checking would need to happen.

keyabarve commented:
Oh okay. So should I call the functions that I wrote in `utils.py` here, or should I basically rewrite the functions here?

ctb commented:
In general, I don't think we should have multiple functions that do the same thing, but I would say to just get something working and tested, and then I can suggest refactorings in review.

keyabarve commented:
Alright, I'll do that. Thanks!

keyabarve committed:
Added check statements and changed a test

keyabarve commented:
@ctb I have commented out the existing check statements that were in the function and also the same statements that were in one of the tests: `tests/test_sourmash_sketch.py::test_do_sourmash_sketchdna_with_bad_scaled`. Please review.

keyabarve commented:
Also, this check: `Dev env instructions / nix (pull_request) Failing after 7m — nix` is failing in all of my PRs and I can't seem to figure out why.

keyabarve committed:
Added scaled tests

keyabarve committed:
Added num tests

keyabarve commented:
I have added the tests for both `num` and `scaled`. Please review.

ctb commented on src/sourmash/command_sketch.py:
```diff
@@ -38,8 +38,14 @@ def _parse_params_str(params_str):
+            #     raise ValueError(f"num is {num}, must be >= 0")
             if num < 0:
-                raise ValueError(f"num is {num}, must be >= 0")
+                raise ValueError(f"ERROR: --num-hashes value must be positive")
```
but --num-hashes isn't used in param strings, so these error messages are problematic, aren't they?

ctb replied:
maybe change the error messages so they refer to num? not sure. "Number of hashes must be >= 0" would maybe be ok.

keyabarve replied:
Yes, I thought that would be a problem. I can just change it to: `ERROR: num value must be positive` if that's okay.

ctb commented on src/sourmash/command_sketch.py:
```diff
@@ -50,10 +56,16 @@ def _parse_params_str(params_str):
-                raise ValueError(f"scaled is {scaled}, must be >= 1")
-            if scaled > 1e8:
-                notify(f"WARNING: scaled value of {scaled} is nonsensical!?")
+                raise ValueError(f"ERROR: --scaled value must be positive")
```
same point here -- --scaled isn't used here so these error messages are misleading.

keyabarve replied:
I'm not sure I understand what you mean. Should it be `scaled` instead of `--scaled` and `num` instead of `--num`?

keyabarve committed:
Changed error statements

keyabarve commented:
I have changed the error statements. Could you take a look at it and let me know if it looks good?

ctb commented:
On Fri, Aug 13, 2021 at 03:03:41PM -0700, Keya Barve wrote:
> Also, this check: `Dev env instructions / nix (pull_request) Failing after 7m ??? nix` is failing in all of my PRs and I can't seem to figure out why.


right, it's not your fault. it's failing on latest too.

ctb commented:
Two thoughts -

it seems like you're just adding new code with new checks, and not centralizing the code so that there is one place for the value checking. In that case, why change the error messages at all? You can just add new ones without changing the old error messages.

That having been said, we will now run into the problem where the code for doing the checking becomes potentially inconsistent.

So I suggest refactoring things so there are two functions, one for each of num and scaled value checking, located in sourmash_args.py. Each function prints out errors or warnings, and may error exit. Then you rework the argparse stuff and the param_string stuff so that those functions are called.

(I'm not quite sure how to get it all to work with argparse properly, though.)

keyabarve commented:
> Two thoughts -
> 
> it seems like you're just adding new code with new checks, and not centralizing the code so that there is one place for the value checking. In that case, why change the error messages at all? You can just add new ones without changing the old error messages.
> 
So then I'll keep the error checking statements for `num` and `scaled` that were already there in place of the code that I wrote and I'll just add the extra statements.

> That having been said, we will now run into the problem where the code for doing the checking becomes potentially inconsistent.
> 
Right, because all the other commands are using the functions from `utils.py` and they will have that specific set of error messages, while the `sketch` command will have different ones.

> So I suggest refactoring things so there are two functions, one for each of num and scaled value checking, located in sourmash_args.py. Each function prints out errors or warnings, and may error exit. Then you rework the argparse stuff and the param_string stuff so that those functions are called.
> 
So then would I have to completely remove the 2 functions from `utils.py` too? And then the same functions that I write in `sourmash_args.py` will be used by all of the commands including `sketch`?

> (I'm not quite sure how to get it all to work with argparse properly, though.)
>
Hmm, I'll take a look into it.

ctb commented:
> > So I suggest refactoring things so there are two functions, one for each of num and scaled value checking, located in sourmash_args.py. Each function prints out errors or warnings, and may error exit. Then you rework the argparse stuff and the param_string stuff so that those functions are called.
> 
> So then would I have to completely remove the 2 functions from `utils.py` too? And then the same functions that I write in `sourmash_args.py` will be used by all of the commands including `sketch`?

Not necessarily - IIRC, the functions in utils.py are argparse-specific, and you could have them call a function in sourmash_args? I think my suggestion would be to get the value checking and error messages about values working in one function (one for num, one for scaled), and then use/display the error messages from that one function in argparse code and in the param_string code, and then we can revisit.

keyabarve commented:
> > > So I suggest refactoring things so there are two functions, one for each of num and scaled value checking, located in sourmash_args.py. Each function prints out errors or warnings, and may error exit. Then you rework the argparse stuff and the param_string stuff so that those functions are called.
> > 
> > 
> > So then would I have to completely remove the 2 functions from `utils.py` too? And then the same functions that I write in `sourmash_args.py` will be used by all of the commands including `sketch`?
> 
> Not necessarily - IIRC, the functions in utils.py are argparse-specific, and you could have them call a function in sourmash_args? I think my suggestion would be to get the value checking and error messages about values working in one function (one for num, one for scaled), and then use/display the error messages from that one function in argparse code and in the param_string code, and then we can revisit.
>
I'm still a little confused, but I'll try something out, push it, and ask for a review.

keyabarve committed:
Rewrote functions and called them and changed error messages

ctb commented:
> I'm still a little confused, but I'll try something out, push it, and ask for a review.

sounds good :)

keyabarve commented:
@ctb I moved the 2 functions from `utils.py` to `sourmash_args.py` and I called them in the `argparse` and `param_string` codes. I also changed the error statements a little, so some of the tests are failing because the error messages don't match. Could you please take a look at it and let me know if it looks good?

keyabarve commented:
@ctb Please review.

ctb commented on tests/test_sourmash_sketch.py:
```diff
@@ -29,6 +29,84 @@
 
 
+def test_do_sourmash_sketch_check_scaled_bounds_negative():
+    with utils.TempDirectory() as location:
```
maybe use runtmp here?

keyabarve replied:
Sure!

ctb commented on src/sourmash/sourmash_args.py:
```diff
@@ -22,11 +22,44 @@
 from . import signature as sigmod
 from .picklist import SignaturePicklist, PickStyle
 from .manifest import CollectionManifest
+import argparse
```
is argparse needed or used below?

keyabarve replied:
Yes, I've used it in these functions: `check_scaled_bounds` and `check_num_bounds` to do this: `raise argparse.ArgumentTypeError`

ctb replied:
oh, ok, I missed that! 👍

ctb commented on src/sourmash/sourmash_args.py:
```diff
@@ -22,11 +22,44 @@
 
 
+def check_scaled_bounds(arg):
+    actual_min_val = 0
```
what is the purpose of defining these here, since they're not used anywhere else?

keyabarve replied:
I'm a little confused. I thought I was supposed to write these functions here in `src/sourmash/sourmash_args.py` and call them in `src/sourmash/command_sketch.py` and `src/sourmash/cli/utils.py`.

ctb replied:
sorry, I meant the constants (`actual_min_val` etc.)

keyabarve replied:
Oh alright, I'll remove them then!

ctb reviewed (commented):
looks like it's heading in the right direction!

keyabarve committed:
Changed sketch tests as requested

keyabarve committed:
Made requested changes

keyabarve commented:
@ctb Please review.

keyabarve committed:
Wrote multigather tests

keyabarve committed:
Deleted changes

keyabarve commented:
@ctb I accidentally wrote a function and pushed it on the current branch (KB_1708) instead of the branch it was actually supposed to be pushed on (KB_1707), but I immediately deleted the function and pushed again. Hence the 2 commits above.

keyabarve commented:
@bluegenes Could you please review this?

bluegenes commented on src/sourmash/cli/utils.py:
```diff
@@ -2,6 +2,8 @@
 import argparse
 from sourmash.logging import notify
+from sourmash.sourmash_args import check_scaled_bounds
+from sourmash.sourmash_args import check_num_bounds
```
```suggestion
from sourmash.sourmash_args import check_scaled_bounds, check_num_bounds
```

keyabarve committed:
Requested changes

Co-authored-by: Tessa Pierce Ward <bluegenes@users.noreply.github.com>

bluegenes commented on src/sourmash/command_sketch.py:
```diff
@@ -7,7 +7,8 @@
                               ComputeParameters)
-
+from sourmash.sourmash_args import check_scaled_bounds
+from sourmash.sourmash_args import check_num_bounds
```
```suggestion
from sourmash.sourmash_args import check_scaled_bounds, check_num_bounds
```

keyabarve committed:
Suggested changes

Co-authored-by: Tessa Pierce Ward <bluegenes@users.noreply.github.com>

bluegenes commented:
This is looking good to me. Why don't you clean up the commented out bits of code and merge in `latest` to see if all tests continue to pass.

keyabarve commented:
Sure! I'll do that and ask for a review. Thanks!

keyabarve committed:
Merge branch 'latest' into KB_1708

keyabarve committed:
Removed comments

keyabarve committed:
Modified error messages in tests

keyabarve commented:
@ctb Should I merge this?

ctb commented:
sure!
On Fri, Aug 27, 2021 at 11:19:31AM -0700, Keya Barve wrote:
> @ctb Should I merge this?
